### PR TITLE
Sb 15165

### PR DESCRIPTION
**Pull request by anmol2302**
<b>Ticket Ref:</b> [SB-15165](https://project-sunbird.atlassian.net/browse/SB-15165)
1: API DOC: [DOC](https://docs.google.com/document/d/1hvfgWTXm4Iw1TBFrJC9ta-LrGI-XJ2dQuTKIqWOX7Zs/edit)
2: Confluence: [DOC](https://project-sunbird.atlassian.net/wiki/spaces/UM/pages/1100414991/Credentials+for+everyone)
3: Certification-service which has following end Points:
    - POST  /certs/v1/add (add data to ES)                          
    - POST  /certs/v1/validate    (validate the data from ES)                 
    - POST  /certs/v1/download  (proxy to the cert-service)             
    - POST  /certs/v1/generate   (proxt to the cert-service)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,419 @@
+## Certification-Service
+
+### Pre-requisites
+0. Make a copy of setVars.sh.sample to list all the environment vars needed.
```
Comment (manzarul) [inline README.md]: Remove step 0 to 3 this is not required now.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+   -H 'Postman-Token: 6a5e0eb0-910a-42d1-9077-c46f6f85397d' \
+   -H 'cache-control: no-cache'`
+
+And, a successful response must be like this:
+
+`{"id":"api.200ok","ver":"v1","ts":"2019-01-17 16:53:26:286+0530","params":{"resmsgid":null,"msgid":"8e27cbf5-e299-43b0-bca7-8347f7ejk5abcf","err":null,"status":"success","errmsg":null},"responseCode":"OK","result":{"response":{"response":"SUCCESS","errors":[]}}}`
+## ElasticSearch V-6.4.0 need to be installed in the system, 
+   - Create Indices and Mappings using below command
+
+## ES-INDICES
```
Comment (manzarul) [inline README.md]: Don't put ES index data inside readme.

File: all-actors/src/main/java/org/sunbird/Application.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * this class is used to instantiate the actor system and open saber.
+ * @author Amit Kumar
+ */
+public class Application {
+    private final static String actorSystemName = "thisActorSystem";
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/Application.java]: chnage actor name here. it should be relevent to code base and same chnages will be applicable under application.conf

File: all-actors/src/main/java/org/sunbird/BaseActor.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    protected Localizer localizer = Localizer.getInstance();
+
+    @Override
+    public void onReceive(Object message) throws Throwable {
+        if (message instanceof Request) {
+            Request request = (Request) message;
+            String operation = request.getOperation();
+            logger.info("BaseActor:onReceive called for operation:" + operation);
+            try {
+                logger.info(String.format("%s:%s:method started at %s",this.getClass().getSimpleName(),operation,System.currentTimeMillis()));
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/BaseActor.java]: String.format is used when we have to pass dynamic value. applicable for all places.

File: all-actors/src/main/java/org/sunbird/CertVars.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    public static String getDOWNLOAD_URI() {
+        return DOWNLOAD_URI;
+    }
+
+
+    private static String getPropsFromEnvs(String props){
+        String propValue=System.getenv(props);
+        if(StringUtils.isBlank(propValue)){
+            logger.error("CertVars:getPropsFromEnvs:no suitable env found for downloadUri");
+            System.exit(-1);
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/CertVars.java]: why it's exit if not found.

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.concurrent.Future;
+
+public class CertsServiceImpl implements ICertService {
+    static Logger logger = Logger.getLogger(CertsServiceImpl.class);
+    private static ObjectMapper requestMapper = new ObjectMapper();
+
+    @Override
+    public String add(Request request) throws BaseException {
+        Map<String, Object> certAddReqMap = new HashMap<>();
+        certAddReqMap = request.getRequest();
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: no need of above line : 
certAddReqMap can be Map<String, Object> certAddReqMap = request.getRequest();

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                    "CertificateActor:addCertificate:provided certificateId exists in record:" + certificatedId);
+            throw new BaseException(IResponseMessage.INVALID_REQUESTED_DATA, IResponseMessage.ID_ALREADY_EXISTS, ResponseCode.CLIENT_ERROR.getCode());
+        }
+        logger.info("CertificateActor:addCertificate:successfully certId not found in records creating new record");
+    }
+
+
+    private void processRecord(Map<String, Object> certReqAddMap){
+        Certificate certificate=getCertificate(certReqAddMap);
+        Map<String,Object>recordMap= requestMapper.convertValue(certificate,Map.class);
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java:62]: why we are first converting map to Certificate object and then certificate object to map back.

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        try {
+            Map<String, Object> certReqMap = new HashMap<>();
+            Map<String,String>requestMap=new HashMap<>();
+            requestMap.put(JsonKeys.PDF_URL,(String)request.getRequest().get(JsonKeys.PDF_URL));
+            certReqMap.put(JsonKeys.REQUEST,requestMap);
+            String requestBody = requestMapper.writeValueAsString(certReqMap);
+            logger.info("CertsServiceImpl:download:request body found:" + requestBody);
+            String apiToCall = CertVars.getSERVICE_BASE_URL().concat(CertVars.getDOWNLOAD_URI());
+            logger.info("CertsServiceImpl:download:complete url found:" + apiToCall);
+            Map<String, String> headerMap = new HashMap<>();
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: Define header map one time as static

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            headerMap.put("Content-Type", "application/json");
+            Future<HttpResponse<JsonNode>>responseFuture=CertificateUtil.makeAsyncPostCall(apiToCall,requestBody,headerMap);
+            HttpResponse<JsonNode> jsonResponse = responseFuture.get();
+            if (jsonResponse != null && jsonResponse.getStatus() == 200) {
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: use httpStatus Code for 200

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                String signedUrl=jsonResponse.getBody().getObject().getJSONObject(JsonKeys.RESULT).getString(JsonKeys.SIGNED_URL);
+                response.put(JsonKeys.SIGNED_URL,signedUrl);
+            } else {
+                throw new BaseException(IResponseMessage.INVALID_REQUESTED_DATA, MessageFormat.format(IResponseMessage.INVALID_PROVIDED_URL,"2222"), ResponseCode.CLIENT_ERROR.getCode());
```
Comment (manzarul) [inline all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: what is "2222" here

File: sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchMapping.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+/** */
+package org.sunbird.helper;
+
+/**
+ * This class will define Elastic search mapping.
+ *
+ * @author Manzarul
+ */
+public class ElasticSearchMapping {
```
Comment (manzarul) [inline sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchMapping.java]: This file is not required.

File: sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchSettings.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+/** */
+package org.sunbird.helper;
+
+/**
+ * This class will define Elastic search default settings.
+ *
+ * @author Manzarul
+ */
+public class ElasticSearchSettings {
```
Comment (manzarul) [inline sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchSettings.java]: this file is not required.

File: sb-es-utils/src/main/resources/elasticsearch.conf [hunk 13]
```diff
@@ -0,0 +1,76 @@
+ {
```
Comment (manzarul) [inline sb-es-utils/src/main/resources/elasticsearch.conf]: this filed is not required now

File: service/conf/application.conf [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+  #
+  #log-config-on-start = true
+}
+
+## Secret key
+# http://www.playframework.com/documentation/latest/ApplicationSecret
+# ~~~~~
+# The secret key is used to sign Play's session cookie.
+# This must be changed for production, but we don't recommend you change it in this file.
+play.http.secret.key = "somekey"
```
Comment (manzarul) [inline service/conf/application.conf]: chnage "somekey" it to some other value.

File: service/conf/routes [hunk 15]
```diff
@@ -0,0 +1,18 @@
+# Routes
+# This file defines all application routes (Higher priority routes first)
+# ~~~~
+
+#Health check
+GET  /health                                  @controllers.health.HealthController.getHealth()
+
+
+# Logs Management APIs
+# POST    /v1.3/system/log/update            @controllers.logsmanager.LogController.setLogLevel()
```
Comment (manzarul) [inline service/conf/routes]: remove this , if not implemented.

File: setVars.sh.sample [hunk 16]
```diff
@@ -0,0 +1,6 @@
+# Sets up all environment vars.
+
+################### CERTIFICATION-SERVICE configuration ###################################
+#export keyA=valA
```
Comment (manzarul) [inline setVars.sh.sample]: Add env if this service required.

File: README.md (hunk 1 above)
Comment (anmol2302) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (anmol2302) [reply README.md]: done

File: all-actors/src/main/java/org/sunbird/Application.java (hunk 3 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/Application.java]: done

File: sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchMapping.java (hunk 11 above)
Comment (anmol2302) [reply sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchMapping.java]: done

File: sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchSettings.java (hunk 12 above)
Comment (anmol2302) [reply sb-es-utils/src/main/java/org/sunbird/helper/ElasticSearchSettings.java]: done

File: service/conf/application.conf (hunk 14 above)
Comment (anmol2302) [reply service/conf/application.conf]: done

File: sb-es-utils/src/main/resources/elasticsearch.conf (hunk 13 above)
Comment (anmol2302) [reply sb-es-utils/src/main/resources/elasticsearch.conf]: done

File: service/conf/routes (hunk 15 above)
Comment (anmol2302) [reply service/conf/routes]: done

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java (hunk 9 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: done

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java (hunk 8 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: done

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java (hunk 7 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java:62]: I personally think to build an object first from request and process it and then send in a form of pure sanitized data map to ES, this increase the readability of code, it just takes less than 2ms to convert back to map.

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java (hunk 10 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: done

File: all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java (hunk 6 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/serviceimpl/CertsServiceImpl.java]: done

File: all-actors/src/main/java/org/sunbird/CertVars.java (hunk 5 above)
Comment (anmol2302) [reply all-actors/src/main/java/org/sunbird/CertVars.java]: done

File: setVars.sh.sample (hunk 16 above)
Comment (anmol2302) [reply setVars.sh.sample]: done

**Review by manzarul — APPROVED** [review]

**Pull request merged**